File: backend/scripts/process_branding.py
#!/usr/bin/env python3
"""
Canonical Branding & Logo Asset Processor for Hayagriva.
Source of Truth: branding/hayagriva_logo.png (User-updated Master Logo)
"""

import os
import base64
import subprocess
import shutil
import re
from PIL import Image, ImageOps

repo_root = "/Users/atulgrover/Desktop/HAYAGRIVA"
master_logo_path = os.path.join(repo_root, "branding/hayagriva_logo.png")
resources_dir = os.path.join(repo_root, "branding/resources")

def process_branding():
    if not os.path.exists(master_logo_path):
        print(f"[ERROR] Master logo file not found at {master_logo_path}")
        return

    print(f"[Branding Processor] Processing master logo: {master_logo_path}...")
    orig_img = Image.open(master_logo_path).convert("RGBA")
    width, height = orig_img.size
    print(f"Master dimensions: {width}x{height}")

    # Copy master logo to backend assets
    os.makedirs(os.path.join(repo_root, "backend/lib/assets"), exist_ok=True)
    orig_img.save(os.path.join(repo_root, "backend/lib/assets/logo.png"))

    # Convert grayscale -> invert luminance for alpha channel
    gray = orig_img.convert("L")
    alpha = ImageOps.invert(gray)

    # 1. Generate white-on-transparent logo (for dark themes & splash loaders)
    white_logo = Image.new("RGBA", (width, height), (255, 255, 255, 0))
    white_logo.putalpha(alpha)
    os.makedirs(resources_dir, exist_ok=True)
    white_logo_path = os.path.join(resources_dir, "logo_white.png")
    white_logo.save(white_logo_path, "PNG")

    # 2. Generate dark-on-transparent logo (for light themes)
    black_logo = Image.new("RGBA", (width, height), (20, 20, 20, 0))
    black_logo.putalpha(alpha)
    black_logo_path = os.path.join(resources_dir, "logo_black.png")
    black_logo.save(black_logo_path, "PNG")

    # Save Base64 text string of white logo for SVG masks & HTML preload
    with open(white_logo_path, "rb") as f:
        base64_white = base64.b64encode(f.read()).decode("utf-8")
    with open(os.path.join(resources_dir, "logo_white_base64.txt"), "w") as f:
        f.write(base64_white)

    # 3. Extract square centered deity icon (512x512)
    side = min(width, height)
    left_offset = (width - side) // 2
    crop_box = (left_offset, 0, left_offset + side, side)

    cropped_alpha = alpha.crop(crop_box).resize((512, 512), Image.Resampling.LANCZOS)
    
    white_icon_512 = Image.new("RGBA", (512, 512), (255, 255, 255, 0))
    white_icon_512.putalpha(cropped_alpha)
    white_icon_512.save(os.path.join(resources_dir, "icon_white_512.png"), "PNG")

    black_icon_512 = Image.new("RGBA", (512, 512), (20, 20, 20, 0))
    black_icon_512.putalpha(cropped_alpha)
    black_icon_512.save(os.path.join(resources_dir, "icon_black_512.png"), "PNG")
    black_icon_512.save(os.path.join(repo_root, "backend/lib/assets/icon.png"), "PNG")

    # 4. Update preload.html
    preload_path = os.path.join(resources_dir, "preload.html")
    if os.path.exists(preload_path):
        with open(preload_path, "r", encoding="utf-8") as f:
            content = f.read()
        content = re.sub(r'<img id="spinner"[\s\S]*?\/>', f'<img id="spinner" src="data:image/png;base64,{base64_white}" alt="Hayagriva Logo" />', content)
        with open(preload_path, "w", encoding="utf-8") as f:
            f.write(content)

    # 5. Update TheiaIDESplash.svg
    splash_svg = f"""<?xml version="1.1" encoding="UTF-8" standalone="no"?>
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 1160 484" width="445.5" height="186">
  <defs>
    <mask id="logo-mask">
      <image href="data:image/png;base64,{base64_white}" x="0" y="0" width="1160" height="484" />
    </mask>
  </defs>
  <rect x="0" y="0" width="1160" height="484" fill="#ffffff" mask="url(#logo-mask)" />
</svg>"""

    splash_path = os.path.join(resources_dir, "TheiaIDESplash.svg")
    with open(splash_path, "w", encoding="utf-8") as f:
        f.write(splash_svg)

    splash_targets = [
        os.path.join(repo_root, "frontend/applications/electron/resources/TheiaIDESplash.svg"),
        os.path.join(repo_root, "frontend/applications/browser/resources/TheiaIDESplash.svg"),
        os.path.join(repo_root, "frontend/applications/electron-next/resources/TheiaIDESplash.svg")
    ]
    for target in splash_targets:
        if os.path.exists(os.path.dirname(target)):
            with open(target, "w", encoding="utf-8") as f:
                f.write(splash_svg)

    # 6. Patch frontend/theia-extensions/hayagriva/src/browser/extension.ts CSS overrides
    with open(os.path.join(resources_dir, "logo_black.png"), "rb") as f:
        base64_black = base64.b64encode(f.read()).decode("utf-8")
    extension_ts_path = os.path.join(repo_root, "frontend/theia-extensions/hayagriva/src/browser/extension.ts")
    if os.path.exists(extension_ts_path):
        with open(extension_ts_path, "r", encoding="utf-8") as f:
            ext_content = f.read()
        target_marker = "svg.theia-WelcomeMessage-Logo {"
        idx = ext_content.find(target_marker)
        if idx != -1:
            bg_marker = "background-image: url('data:image/png;base64,"
            bg_idx = ext_content.find(bg_marker, idx)
            if bg_idx != -1:
                start_data = bg_idx + len(bg_marker)
                end_data = ext_content.find("');", start_data)
                if end_data != -1:
                    new_ext_content = ext_content[:start_data] + base64_black + ext_content[end_data:]
                    with open(extension_ts_path, "w", encoding="utf-8") as f:
                        f.write(new_ext_content)
                    print("Updated base64 dark logo in extension.ts WelcomeMessage-Logo CSS override!")

    # 7. Synchronize product extension icons
    product_icons_dir = os.path.join(repo_root, "frontend/theia-extensions/product/src/browser/icons")
    if os.path.exists(product_icons_dir):
        black_logo.save(os.path.join(product_icons_dir, "TheiaIDE.png"), "PNG")
        white_logo.save(os.path.join(product_icons_dir, "TheiaIDE-next.png"), "PNG")
        black_icon_512.save(os.path.join(product_icons_dir, "512-512.png"), "PNG")
        white_icon_512.save(os.path.join(product_icons_dir, "512-512-next.png"), "PNG")

    icon_destinations = [
        os.path.join(repo_root, "frontend/applications/electron/resources/icon.png"),
        os.path.join(repo_root, "frontend/applications/browser/resources/icon.png"),
        os.path.join(resources_dir, "icons/512x512.png"),
        os.path.join(resources_dir, "icons/WindowIcon/512-512.png"),
        os.path.join(resources_dir, "icons/LinuxLauncherIcons/512x512.png"),
        os.path.join(repo_root, "frontend/applications/electron/resources/icons/MacLauncherIcons/icon.icon/Assets/icon.png"),
        os.path.join(resources_dir, "icons/MacLauncherIcons/icon.icon/Assets/icon.png"),
    ]
    for dest in icon_destinations:
        os.makedirs(os.path.dirname(dest), exist_ok=True)
        black_icon_512.save(dest, "PNG")

    # Generate JPG assets for Theia Light BG / Dark BG
    jpg_bg_light = Image.new("RGB", (512, 512), (255, 255, 255))
    jpg_bg_light.paste(black_icon_512, (0, 0), black_icon_512)
    jpg_bg_dark = Image.new("RGB", (512, 512), (30, 30, 30))
    jpg_bg_dark.paste(white_icon_512, (0, 0), white_icon_512)

    jpg_destinations = [
        (jpg_bg_light, os.path.join(repo_root, "frontend/applications/electron/resources/icons/MacLauncherIcons/icon.icon/Assets/Theia Light BG.jpg")),
        (jpg_bg_dark, os.path.join(repo_root, "frontend/applications/electron/resources/icons/MacLauncherIcons/icon.icon/Assets/Theia Dark BG.jpg")),
        (jpg_bg_light, os.path.join(resources_dir, "icons/MacLauncherIcons/icon.icon/Assets/Theia Light BG.jpg")),
        (jpg_bg_dark, os.path.join(resources_dir, "icons/MacLauncherIcons/icon.icon/Assets/Theia Dark BG.jpg")),
    ]
    for img, dest in jpg_destinations:
        os.makedirs(os.path.dirname(dest), exist_ok=True)
        img.save(dest, "JPEG")

    # 8. Generate native macOS .icns icons
    tmp_iconset = os.path.join(resources_dir, ".hayagriva_tmp.iconset")
    if os.path.exists(tmp_iconset):
        shutil.rmtree(tmp_iconset)
    os.makedirs(tmp_iconset, exist_ok=True)

    sizes = [16, 32, 64, 128, 256, 512]
    for sz in sizes:
        resized = black_icon_512.resize((sz, sz), Image.Resampling.LANCZOS)
        resized.save(os.path.join(tmp_iconset, f"icon_{sz}x{sz}.png"), "PNG")
        resized_2x = black_icon_512.resize((sz * 2, sz * 2), Image.Resampling.LANCZOS)
        resized_2x.save(os.path.join(tmp_iconset, f"icon_{sz}x{sz}@2x.png"), "PNG")

    icns_output_path = os.path.join(repo_root, "Hayagriva.app/Contents/Resources/hayagriva.icns")
    electron_icns = os.path.join(repo_root, "frontend/applications/electron/resources/icon.icns")
    electron_mac_icns = os.path.join(repo_root, "frontend/applications/electron/resources/icons/MacLauncherIcons/icon.icns")
    browser_icns = os.path.join(repo_root, "frontend/applications/browser/resources/icon.icns")
    branding_mac_icns = os.path.join(resources_dir, "icons/MacLauncherIcons/icon.icns")
    branding_icns = os.path.join(resources_dir, "icon.icns")

    try:
        subprocess.run(["iconutil", "-c", "icns", tmp_iconset, "-o", icns_output_path], check=True)
        shutil.copyfile(icns_output_path, electron_icns)
        shutil.copyfile(icns_output_path, electron_mac_icns)
        shutil.copyfile(icns_output_path, browser_icns)
        shutil.copyfile(icns_output_path, branding_mac_icns)
        shutil.copyfile(icns_output_path, branding_icns)
        print(f"Generated macOS .icns icon file at {icns_output_path}")
    except Exception as e:
        print(f"iconutil note: {e}")
    finally:
        shutil.rmtree(tmp_iconset, ignore_errors=True)

    print("✓ All branding assets successfully compiled from master logo!")

if __name__ == "__main__":
    process_branding()

File: backend/scripts/process_branding.js
#!/usr/bin/env node
/**
 * process_branding.js
 * Canonical Node.js Branding & Logo Asset Processor for Hayagriva.
 *
 * Reads the master logo from branding/hayagriva_logo.png and generates:
 *   - Transparent PNG banners for UI panels (TheiaIDE.png / TheiaIDE-next.png) so they blend seamlessly with any background theme.
 *   - A solid white-background 512x512 square icon for window/dock/launcher use.
 *   - All destination copies across frontend/, branding/, backend/.
 *   - macOS .icns via iconutil (macOS only).
 *
 * Pure Node.js — uses @napi-rs/canvas (already installed in backend/).
 *
 * Usage: node backend/scripts/process_branding.js
 */

'use strict';

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');
const { createCanvas, loadImage } = require('@napi-rs/canvas');

// ── Path constants ────────────────────────────────────────────────────────────
const repoRoot = path.resolve(__dirname, '../..');
const masterLogoPath = path.join(repoRoot, 'branding/hayagriva_logo.png');
const resourcesDir = path.join(repoRoot, 'branding/resources');

// ── Helper: save canvas to PNG ────────────────────────────────────────────────
function saveCanvas(canvas, dest) {
    fs.mkdirSync(path.dirname(dest), { recursive: true });
    const buf = canvas.toBuffer('image/png');
    fs.writeFileSync(dest, buf);
}

// ── Generate square icon (white bg, logo centred with padding) ───────────────
async function makeSquareIcon(masterImg, size = 512, padding = 40) {
    const canvas = createCanvas(size, size);
    const ctx = canvas.getContext('2d');

    // White background for app launcher icon
    ctx.fillStyle = '#ffffff';
    ctx.fillRect(0, 0, size, size);

    // Scale logo to fit inside padded area, maintaining aspect ratio
    const available = size - padding * 2;
    const scale = Math.min(available / masterImg.width, available / masterImg.height);
    const dw = Math.round(masterImg.width * scale);
    const dh = Math.round(masterImg.height * scale);
    const dx = Math.round((size - dw) / 2);
    const dy = Math.round((size - dh) / 2);

    ctx.drawImage(masterImg, dx, dy, dw, dh);
    return canvas;
}

// ── Generate transparent banner for UI (light or dark theme variant) ───────────
async function makeTransparentBanner(masterImg, bw = 600, bh = 300, padH = 20, padV = 20, variant = 'black') {
    const canvas = createCanvas(bw, bh);
    const ctx = canvas.getContext('2d');

    // Scale logo to fit banner area
    const scale = Math.min((bw - padH * 2) / masterImg.width, (bh - padV * 2) / masterImg.height);
    const dw = Math.round(masterImg.width * scale);
    const dh = Math.round(masterImg.height * scale);
    const dx = Math.round((bw - dw) / 2);
    const dy = Math.round((bh - dh) / 2);

    ctx.drawImage(masterImg, dx, dy, dw, dh);

    // Manipulate pixels to make white background transparent
    const imgData = ctx.getImageData(0, 0, bw, bh);
    const data = imgData.data;

    for (let i = 0; i < data.length; i += 4) {
        const r = data[i];
        const g = data[i + 1];
        const b = data[i + 2];
        const a = data[i + 3];

        if (a === 0 || (r > 235 && g > 235 && b > 235)) {
            // White or already transparent background
            data[i + 3] = 0;
        } else if (variant === 'white') {
            // Dark theme: turn artwork pixels into white
            data[i] = 255;
            data[i + 1] = 255;
            data[i + 2] = 255;
        }
    }

    ctx.putImageData(imgData, 0, 0);
    return canvas;
}

// ── Main ──────────────────────────────────────────────────────────────────────
async function processBranding() {
    console.log('[Branding Processor] Reading master logo:', masterLogoPath);

    if (!fs.existsSync(masterLogoPath)) {
        console.error('[ERROR] Master logo not found:', masterLogoPath);
        process.exit(1);
    }

    const masterImg = await loadImage(masterLogoPath);
    console.log(`[Branding Processor] Master dimensions: ${masterImg.width}x${masterImg.height}`);

    // ── Generate assets ───────────────────────────────────────────────────────
    const squareCanvas = await makeSquareIcon(masterImg, 512, 40);
    const lightBannerCanvas = await makeTransparentBanner(masterImg, 600, 300, 20, 20, 'black');
    const darkBannerCanvas = await makeTransparentBanner(masterImg, 600, 300, 20, 20, 'white');

    // ── 1. Window / Dock icon destinations (512x512 square) ──────────────────
    const squareDestinations = [
        'frontend/applications/electron/resources/icons/WindowIcon/512-512.png',
        'frontend/applications/electron/resources/icon.png',
        'frontend/applications/electron/resources/icons/MacLauncherIcons/icon.icon/Assets/icon.png',
        'frontend/applications/electron/resources/icons/LinuxLauncherIcons/512x512.png',
        'frontend/applications/electron/resources/icons/512x512.png',
        'frontend/applications/browser/resources/icon.png',
        'branding/resources/icons/WindowIcon/512-512.png',
        'branding/resources/icons/LinuxLauncherIcons/512x512.png',
        'branding/resources/icons/512x512.png',
        'branding/resources/icons/MacLauncherIcons/icon.icon/Assets/icon.png',
        'branding/resources/icon_black_512.png',
        'branding/resources/icon_white_512.png',
    ];

    for (const rel of squareDestinations) {
        saveCanvas(squareCanvas, path.join(repoRoot, rel));
    }
    console.log('✓ Synchronized square icon PNGs across all destinations');

    // ── 2. Product extension banners (Transparent backgrounds) ─────────────
    saveCanvas(lightBannerCanvas, path.join(repoRoot, 'frontend/theia-extensions/product/src/browser/icons/TheiaIDE.png'));
    saveCanvas(darkBannerCanvas, path.join(repoRoot, 'frontend/theia-extensions/product/src/browser/icons/TheiaIDE-next.png'));
    saveCanvas(lightBannerCanvas, path.join(repoRoot, 'branding/resources/logo_black.png'));
    saveCanvas(darkBannerCanvas, path.join(repoRoot, 'branding/resources/logo_white.png'));
    saveCanvas(lightBannerCanvas, path.join(repoRoot, 'docs/marketing/resources/hayagriva_logo_black.png'));
    saveCanvas(darkBannerCanvas, path.join(repoRoot, 'docs/marketing/resources/hayagriva_logo_white.png'));

    // Also copy 512x512 to product icons folder
    saveCanvas(squareCanvas, path.join(repoRoot, 'frontend/theia-extensions/product/src/browser/icons/512-512.png'));
    saveCanvas(squareCanvas, path.join(repoRoot, 'frontend/theia-extensions/product/src/browser/icons/512-512-next.png'));
    console.log('✓ Synchronized transparent product extension banners and icon PNGs');

    // ── 3. Backend assets ─────────────────────────────────────────────────────
    const backendAssetsDir = path.join(repoRoot, 'backend/lib/assets');
    fs.mkdirSync(backendAssetsDir, { recursive: true });
    fs.copyFileSync(masterLogoPath, path.join(backendAssetsDir, 'logo.png'));
    saveCanvas(squareCanvas, path.join(backendAssetsDir, 'icon.png'));
    console.log('✓ Synchronized backend assets');

    // ── 4. macOS .icns generation via iconutil (macOS only) ──────────────────
    const tmpIconset = path.join(resourcesDir, '.hayagriva_tmp.iconset');
    if (fs.existsSync(tmpIconset)) fs.rmSync(tmpIconset, { recursive: true });
    fs.mkdirSync(tmpIconset, { recursive: true });

    const sizes = [16, 32, 64, 128, 256, 512];
    for (const sz of sizes) {
        const c = await makeSquareIcon(masterImg, sz, Math.max(2, Math.round(sz * 0.08)));
        saveCanvas(c, path.join(tmpIconset, `icon_${sz}x${sz}.png`));
        const c2x = await makeSquareIcon(masterImg, sz * 2, Math.max(4, Math.round(sz * 0.16)));
        saveCanvas(c2x, path.join(tmpIconset, `icon_${sz}x${sz}@2x.png`));
    }

    const icnsOut = path.join(repoRoot, 'Hayagriva.app/Contents/Resources/hayagriva.icns');
    try {
        fs.mkdirSync(path.dirname(icnsOut), { recursive: true });
        execSync(`iconutil -c icns "${tmpIconset}" -o "${icnsOut}"`, { stdio: 'pipe' });

        const icnsDestinations = [
            'frontend/applications/electron/resources/icon.icns',
            'frontend/applications/electron/resources/icons/MacLauncherIcons/icon.icns',
            'frontend/applications/browser/resources/icon.icns',
            'branding/resources/icons/MacLauncherIcons/icon.icns',
            'branding/resources/icon.icns',
        ];
        for (const rel of icnsDestinations) {
            const dest = path.join(repoRoot, rel);
            fs.mkdirSync(path.dirname(dest), { recursive: true });
            fs.copyFileSync(icnsOut, dest);
        }
        console.log('✓ Generated macOS .icns icons');
    } catch (e) {
        console.warn('iconutil note (non-fatal):', e.message);
    } finally {
        fs.rmSync(tmpIconset, { recursive: true, force: true });
    }

    console.log('✓ All branding assets successfully compiled from master logo with transparent UI banners!');
}

if (require.main === module) {
    processBranding().catch(err => {
        console.error('[Branding Processor] Fatal error:', err);
        process.exit(1);
    });
}

module.exports = { processBranding };
